### PR TITLE
VIDEO-4517: Add Client Track Switch Off Control Support

### PR DESCRIPTION
**Pull request by ceaglest**
This PR adds support for Client Track Switch Off Control to the iOS Video App.

* Client track switch off control is unset by default (resolves to auto)
* Added feature to settings UI / settings store
* The app does not implement any specific manual controls. Selecting manual is effectively the legacy behavior without a `maxTracks` limit
* Removed the deprecated `maxTracks` property.
* Removed extra UICollectionView padding on the leading edge.
* CollectionView logic was updated to add and remove renderers at cell display time rather than cell dequeue time (thanks @timrozum)

Note: While testing this feature I ran into a background CPU usage crash which was resolved in https://github.com/twilio/twilio-video-app-ios/pull/155.

**Contributing to Twilio**

> All third-party contributors acknowledge that any contributions they provide will be made under the same open-source license that the open-source project is provided under.

- [x] I acknowledge that all my contributions will be made under the project's license.

### CODE REVIEW

File: VideoApp/VideoApp/Video/Room/ConnectOptionsFactory.swift [hunk 1]
```diff
@@ -49,14 +49,15 @@ import TwilioVideo
                 videoOptions: VideoBandwidthProfileOptions { builder in
                     builder.mode = TwilioVideo.BandwidthProfileMode(setting: self.appSettingsStore.bandwidthProfileMode)
                     builder.maxSubscriptionBitrate = self.appSettingsStore.maxSubscriptionBitrate as NSNumber?
-                    builder.maxTracks = self.appSettingsStore.maxTracks as NSNumber?
+                    // builder.maxTracks = self.appSettingsStore.maxTracks as NSNumber?
```
Comment (ceaglest) [inline VideoApp/VideoApp/Video/Room/ConnectOptionsFactory.swift:52]: Should this app demonstrate deprecated APIs?

File: VideoApp/VideoApp/Stores/AppSettings/SubscribedTrackSwitchOffMode.swift [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+import Foundation
+
+enum SubscribedTrackSwitchOffMode: String, SettingOptions {
+    case sdkDefault
+    case auto
+    case manual
+    
+    var title: String {
+        switch self {
+        case .sdkDefault: return "SDK Default (Auto)"
```
Comment (timrozum) [inline VideoApp/VideoApp/Stores/AppSettings/SubscribedTrackSwitchOffMode.swift:26]: For `Server Default` we don't specify what the default actually is so maybe we shouldn't for SDK either. Perhaps unlikely to cause a problem but it is out of the app's control in any case.

**Review by timrozum — COMMENTED** [review]
This looks good so far. 

I wouldn't really be concerned about tests at this point because most of this is settings which we don't have tests for and I expect will be replaced by `InAppSettingsKit` before long.

When the time comes we could consider unit tests for `ConnectOptionsFactory` but I can take care of that.

UI tests are probably not practical right now for this feature.

Happy to help if the cell reuse stuff needs work.

File: VideoApp/VideoApp/Stores/AppSettings/SubscribedTrackSwitchOffMode.swift (hunk 2 above)
Comment (ceaglest) [reply VideoApp/VideoApp/Stores/AppSettings/SubscribedTrackSwitchOffMode.swift]: Updated.

File: VideoApp/VideoApp/Storyboards/Base.lproj/Main.storyboard [hunk 3]
```diff
@@ -528,7 +528,7 @@
                                         </constraints>
                                     </view>
                                     <collectionView clipsSubviews="YES" multipleTouchEnabled="YES" contentMode="scaleToFill" showsHorizontalScrollIndicator="NO" showsVerticalScrollIndicator="NO" dataMode="prototypes" translatesAutoresizingMaskIntoConstraints="NO" id="wU6-sZ-Ash">
-                                        <rect key="frame" x="10" y="690" width="404" height="172"/>
+                                        <rect key="frame" x="0.0" y="690" width="414" height="172"/>
```
Comment (ceaglest) [inline VideoApp/VideoApp/Storyboards/Base.lproj/Main.storyboard:531]: The UICollectionView had a gap on the left side of 10 points on the leading edge but was flush to the trailing edge. I think this was just an unintended bug (it's not obvious until you have more than a screen full of cells).

**Comment by ceaglest** [comment]
Hi @timrozum,

I continued to test this PR yesterday. What I noticed was that UICollectionViewDataSource prefetching is preventing some of the subscribed track switch off optimizations. Specifically, the behavior where cells can be opportunistically dequeued by UIKit before they come onscreen and are not recycled immediately after they go offscreen.

I think the best way to manage rendering in the UICollectionView is to implement the delegate methods for cell display. I was wondering if you would be interested in taking a shot at it?

- `func collectionView(_ collectionView: UICollectionView, willDisplay cell: UICollectionViewCell, forItemAt indexPath: IndexPath)`
- `func collectionView(_ collectionView: UICollectionView, didEndDisplaying cell: UICollectionViewCell, forItemAt indexPath: IndexPath)`

We should make sure that video starts / stops rendering within these methods. To me it is okay to keep the existing dequeue logic (`cellForItemAtIndexPath`) as an optimization to have a better chance that video will be ready by the time the cell is drawn to screen.

**Comment by timrozum** [comment]
@ceaglest yeah I had looked at `didEndDisplaying` yesterday and also thought that would be useful for this. 

I guess `willDisplay` may also be necessary if it is possible for a cell to be in use but not visible from collection view perspective. But haven't looked into that.

**Comment by ceaglest** [comment]
> I guess willDisplay may also be necessary if it is possible for a cell to be in use but not visible from collection view perspective. But haven't looked into that.

Yes it is needed as well. To replicate, try scrolling a cell just offscreen and then back on. Typically the cell does not go through the reuse queue during this action but your delegate receives a `didEndDisplaying` followed by a `willDisplay`.

**Comment by timrozum** [comment]
Thanks great to know.

File: Podfile [hunk 4]
```diff
@@ -1,3 +1,4 @@
+source 'git@github.com:twilio/cocoapod-specs-internal.git'
```
Comment (timrozum) [inline Podfile:1]: Remove before merging.

**Comment by ceaglest** [comment]
I am curious if we would cut a feature branch for media optimizations in advance of the 4.5.0 release? This app should also be updated for VideoContentPreferences, which I think deserves a separate PR.

File: VideoApp/VideoApp/Storyboards/Base.lproj/Main.storyboard (hunk 3 above)
Comment (timrozum) [reply VideoApp/VideoApp/Storyboards/Base.lproj/Main.storyboard:531]: I think I set this gap intentionally but not sure what desired UX really is to be honest. I'm good with this adjustment. 

Probably how it should really work someday is the collection view has no gaps on either side and instead we use spacing inside of the collection view content at start and end. This way it looks good at start/end position and when scrolling (cells are not clipped before the edge of the display). 

But I think the team wants to completely redo UX of this screen someday so best to tackle then.

**Review by ceaglest — COMMENTED** [review]
Haven't run the changes locally but it looks like they will resolve the flickering issue. Thank you @timrozum. 👍

**Review by timrozum — APPROVED** [review]

**Pull request merged**